src/balancer-strategy-manager/BalancerV2WeightedPoolStrategyManager.sol

```diff
@@ -0,0 +1,226 @@
+    }
+    _;
+  }
+
+  constructor(
+    address _vault,
+    bytes32 _poolId,
+    TokenConfig[] memory _tokenConfig,
+    address _owner,
+    address _guardian,
+    address _hypernative
+  ) {
+    VAULT = IBalancerVault(_vault);
+    POOL_ID = _poolId;
+
+    (address poolAddress, ) = VAULT.getPool(_poolId);
+    POOL = IBalancerPool(poolAddress);
+
+    address[] memory actuallTokens;
+    (actuallTokens, , ) = VAULT.getPoolTokens(_poolId);
+
+    TOKEN_COUNT = actuallTokens.length;
+
+    if (_tokenConfig.length != TOKEN_COUNT) {
+      revert TokenCountMismatch();
+    }
+
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      if (actuallTokens[i] != _tokenConfig[i].token) {
+        revert TokenMismatch();
+      }
+
+      tokenConfig[i] = _tokenConfig[i];
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    _transferOwnership(_owner);
+    _updateGuardian(_guardian);
+    HYPERNATIVE = _hypernative;
+  }
+
+  /**
+   * @dev sets token provider
+   * @param _id The index of token to set
+   * @param _provider The address of provider
+   */
+  function setTokenProvider(uint256 _id, address _provider) external onlyOwner {
+    if (_id >= TOKEN_COUNT) {
+      revert TokenCountMismatch();
+    }
+
+    tokenConfig[_id].provider = _provider;
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function deposit(uint256[] calldata _tokenAmounts) external onlyOwnerOrGuardian {
+    if (_tokenAmounts.length != TOKEN_COUNT) {
+      revert TokenCountMismatch();
+    }
+
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      IERC20 token = IERC20(tokenConfig[i].token);
+
+      if (_tokenAmounts[i] > token.balanceOf(address(this))) {
+        revert InsufficientToken(address(token));
+      }
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    bytes memory userData = abi.encode(WeightedPoolUserData.JoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT);
+
+    IAsset[] memory assets = new IAsset[](TOKEN_COUNT);
+    for (uint256 i = 0; i < TOKEN_COUNT; ) {
+      IERC20(tokenConfig[i].token).safeIncreaseAllowance(address(VAULT), _tokenAmounts[i]);
+      assets[i] = IAsset(tokenConfig[i].token);
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    userData = abi.encode(
+      WeightedPoolUserData.JoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT,
+      _tokenAmounts,
+      0
+    );
+
+    IBalancerVault.JoinPoolRequest memory request = IBalancerVault.JoinPoolRequest({
+      assets: assets,
+      maxAmountsIn: _tokenAmounts,
+      userData: userData,
+      fromInternalBalance: false
+    });
+
+    VAULT.joinPool(POOL_ID, address(this), address(this), request);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function withdraw(uint256 bpt) external onlyOwnerOrGuardian returns (uint256[] memory) {
+    return _withdraw(bpt);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function emergencyWithdraw() external onlyWithdrawable returns (uint256[] memory) {
+    uint256 bptAmount = IERC20(address(POOL)).balanceOf(address(this));
```

Comment:
No need to set this as a variable, which consumes gas, can just pass the IERC.balanceOf